contributors/design-proposals/storage/volume-topology-scheduling.md

```diff
@@ -266,41 +304,368 @@ const (
+to the provisioner.
+
+TODO: would this be better represented as part of StorageClass quota since that
+is where admins currently specify policies?
+However, ResourceQuota currently only supports quantity, and topology
+restrictions have to be evaluated during scheduling, not admission.
+
+#### Zonal Example
+This example restricts the volumes provisioned to zones us-central1-a and
+us-central1-b.
+```
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  Name: zonal-class
+provisioner: kubernetes.io/gce-pd
+parameters:
+  type: pd-standard
+allowedTopologies:
+  requiredToplogies:
+  - matchNodeLabels:
+      failure-domain.beta.kubernetes.io/zone: us-central1-a
+  - matchNodeLabels:
+      failure-domain.beta.kubernetes.io/zone: us-central1-b
+```
+
+#### Multi-Zonal Example
+This example restricts the volume's primary and failover zones
+to us-central1-a and us-central1-b. Topologies that are incompatible with the
+storage provider parameters will be enforced by the provisioner.
+```
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  Name: multi-zonal-class
+provisioner: kubernetes.io/gce-pd
+parameters:
+  type: pd-standard
+  replicas: 2
+allowedTopologies:
+  requiredToplogies:
+  - matchNodeLabels:
+      failure-domain.beta.kubernetes.io/zone: us-central1-a
+  - matchNodeLabels:
+      failure-domain.beta.kubernetes.io/zone: us-central1-b
+```
+
+#### Multi Label Example
+This example restricts the volume's topology to nodes that
+have the following labels:
+
+* "zone: us-central1-a" and "rack: rack1" or,
+* "zone: us-central1-b" and "rack: rack1" or,
+* "zone: us-central1-b" and "rack: rack2"
+
+```
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  Name: something-fancy
+provisioner: rack-based-provisioner
+parameters:
+allowedTopologies:
+  requiredTopologies:
+  - matchNodeLabels:
+      zone: us-central1-a
+      rack: rack1
+  - matchNodeLabels:
+      zone: us-central1-b
+      rack: rack1
+  - matchNodeLabels:
+      zone: us-central1-b
+      rack: rack2
+```
+
+### Storage Provider Capacity Reporting
+Some volume types have capacity limitations per topology domain. For example,
+local volumes have limited capacity per node. Provisioners for these volume
+types can report these capacity limitations through a new StorageClass status
+field.
+
+```
+type StorageClass struct {
+    ...
+
+    Status *StorageClassStatus
+}
+
+type StorageClassStatus struct {
+    Capacity *StorageClassCapacity
+}
+
+type StorageClassCapacity struct {
+    // The node label key that the provisioner operates on
+    TopologyKey string
+    // Capacities specifies the total capacity available
+    // (free and used) for provisioning at each value of the
+    // TopologyKey
+    Capacities map[string]resource.Quantity
+}
+```
+
+If no StorageClassCapacity is specified, then there are no capacity
+limitations.
+
+Capacity reporting is limited to a single topology key. `TopologyKey`
```

Comment:
I think it doesn't have to be restricted to a single key if the Capacities map key is a VolumeTopology type.  Not sure if that is expressible in the API though, I haven't seen any non-string map keys in the API.

If we can't do the above, then Capacities could be an array, but then lookups by the scheduler will be inefficient.  We need to figure out how important capacity reporting is for multi-dimensional storage systems.  The examples I can think of are only hypothetical at this point.

Reply:
This would indeed bring inconvenience to provisioners and the scheduler,
I would suggest we do not introduce this before we can see practical use case of multi-dimensional  capacity reporting

Reply:
We can support multi-keys but still have efficient lookups by flattening the values into a single string separated by a delimiter.  Will update the spec shortly.

Reply:
Also there is a limitation that a single storage pool should not be shared by multiple storage classes, otherwise you will have more capacity reported than is actually available.